Speed up PHP-fpm container  on MacOS

tonicospinelli opened the pull request:
Many MacOS users may be facing slow responses from Shop or Admin when
running the sylius application in the container. The main reason is how
docker was build in the MacOS architecture. The PHP application has
thousands of files to be loaded in memory (vendor directory mainly).

So, to decrease the page response time, the vendor directory is not
synced into the container. It will be ignored by docker-composer.yml
using the named volume.

Also, was added a health check for the PHP-fpm container to know when it
is ready. If you want to check healthiness, just run the command
`docker-compose ps` to see the container State.

The HEALTHCHECK instruction tells Docker how to test a container
to check that it is still working. This can detect cases such as a
web server stuck in an infinite loop and unable to handle new
connections, even though the server process is still running.

The PHP and PHP-FPM configurations received some adjustments to get
better support development environment experience using containers.

I did pick up these updates from two other contributors that I
co-authored in this commit.

Co-authored-by: Kévin Dunglas <dunglas@gmail.com>
Co-authored-by: arti0090 <arti0090@gmail.com>

tonicospinelli committed:
Speed up PHP-fpm container on MacOS

Many MacOS users may be facing slow responses from Shop or Admin when
running the sylius application in the container. The main reason is how
docker was build in the MacOS architecture. The PHP application has
thousands of files to be loaded in memory (vendor directory mainly).

So, to decrease the page response time, the vendor directory is not
synced into the container. It will be ignored by docker-composer.yml
using the named volume.

Also, was added a health check for the PHP-fpm container to know when it
is ready. If you want to check healthiness, just run the command
`docker-compose ps` to see the container State.

The HEALTHCHECK instruction tells Docker how to test a container
to check that it is still working. This can detect cases such as a
web server stuck in an infinite loop and unable to handle new
connections, even though the server process is still running.

The PHP and PHP-FPM configurations received some adjustments to get
better support development environment experience using containers.

I did pick up these updates from two other contributors that I
co-authored in this commit.

Co-authored-by: Kévin Dunglas <dunglas@gmail.com>
Co-authored-by: arti0090 <arti0090@gmail.com>

tonicospinelli commented:
In the environment with `APP_DEV = dev` and ` APP_DEBUG = 1`, the response time is around ~2s or more.
As for `APP_DEV = prod` and ` APP_DEBUG = 0`, the response time is about ~300ms or more.

Of course, it depends on your workstation and the resources available for the containers. Therefore, I got a better experience in the local environment with these changes.

Always the first request took more time than I mentioned above, it is because the PHP is caching code on opcache and Symfony generating the cache files.

vvasiloi commented on .env.prod:
```diff
@@ -1,2 +1,3 @@
+APP_ENV=prod
```
This file is only loaded when the env = prod, so it's redundant.

tonicospinelli replied:
@vvasiloi I forgot to remove 🤗

vvasiloi commented on docker-compose.yml:
```diff
@@ -28,13 +28,15 @@ services:
-      - MYSQL_DATABASE=sylius
-      - MYSQL_USER=sylius
+      - MYSQL_DATABASE=sylius_${APP_ENV}
+      - MYSQL_USER=${MYSQL_ROOT_PASSWORD:-sylius}
```
Unedited copy/paste?

tonicospinelli replied:
Nice catch! my fault to not review it before pull

tonicospinelli committed:
Load dotenv into docker composer build args

The container image was ignoring the dotenv config to build the PHP
container image.

tonicospinelli commented:
@vvasiloi I fix the typos 😊

lchrusciel commented:
@Arminek can you take a look into this?

Ferror commented:
It's a nice improvement, but with the latest Docker Desktop filesystem, there is no need for performance improvement.

https://www.docker.com/blog/speed-boost-achievement-unlocked-on-docker-desktop-4-6-for-mac/

I don't mind keeping the health checks in the environment :)

jacquesbh commented:
@Ferror Docker for mac still has some issues unresolved :). But soonish I hope!

rust-le commented:
I'm closing this, if necessary, please create a separate topic
